### PR TITLE
#4668 Persist unsaved changes in page editor

### PR DESCRIPTION
**Pull request by BLoe**
## What does this PR do?

- Closes #4668 

## Demo

https://www.loom.com/share/75d10b8d2aaf4206a3412ae9c269b14d

## QA

[See _slice 1_ on this page](https://www.notion.so/pixiebrix/Persist-Unsaved-Changes-aea30324fd574a54b481b27c18087973)

## Checklist

- [ ] Add tests
- [ ] Run Storybook and manually confirm that all stories are working
- [x] Designate a primary reviewer

### CODE REVIEW

File: src/auth/authSlice.ts [hunk 1]
```diff
@@ -41,11 +42,14 @@ export const authSlice = createSlice({
   },
 });
 
+// Shadowing this type in order to export it properly
```
Comment (twschiller) [inline src/auth/authSlice.ts:45]: Could you clarify the code comment? Not sure what the "it" refers to

File: src/pageEditor/slices/editorSlice.ts [hunk 2]
```diff
@@ -878,3 +880,11 @@ export const actions = {
   checkAvailableDynamicElements,
   checkActiveElementAvailability,
 };
+
+// Shadowing this type in order to export it properly
```
Comment (twschiller) [inline src/pageEditor/slices/editorSlice.ts:884]: Same comment as here: https://github.com/pixiebrix/pixiebrix-extension/pull/4734/files#r1038603170

File: src/store/StorageInterface.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+// Exporting from redux-persist-webextension-storage to fix the
+// "exporting variable using name but cannot be named" lint errors
+export interface StorageInterface {
+  getItem(key: string): Promise<any>;
+  removeItem(key: string): Promise<void>;
+  setItem(key: string, value: any): Promise<void>;
```
Comment (twschiller) [inline src/store/StorageInterface.ts:23]: Ideally `value` type would be unknown

File: src/auth/authSlice.ts [hunk 4]
```diff
@@ -41,11 +42,14 @@ export const authSlice = createSlice({
...
+const local: StorageInterface = localStorage;
+
 // TODO refactor to use token.ts/updateUserData
 // Current approach is not ideal, AuthState is cached along with UserData (which is used by background script).
 export const persistAuthConfig = {
   key: "authOptions",
-  storage: localStorage,
+  storage: local,
```
Comment (twschiller) [inline src/auth/authSlice.ts:52]: Why not just do `local as StorageInterface here?

File: src/recipes/recipesSlice.ts [hunk 5]
```diff
@@ -110,3 +112,11 @@ export const recipesActions = {
   loadRecipesFromCache,
   refreshRecipes,
 };
+
+// Shadowing this type in order to export it properly
+const local: StorageInterface = localStorage;
+
+export const persistRecipesConfig = {
```
Comment (twschiller) [inline src/recipes/recipesSlice.ts:119]: Why are we persisting this? It gets loaded from the cache

**Comment by twschiller** [comment]
Other comments:
- If we make this change, we'll need to define migrations for all future editor slice state shape changes
- It would be good to draw attention to that by explicitly setting the "version" to 1: https://github.com/pixiebrix/pixiebrix-extension/blob/b6c2ba9c662c4db12af27c5a611c843285e5c6e6/src/store/extensionsStorage.ts#L70-L70

File: src/auth/authSlice.ts (hunk 4 above)
Comment (BALEHOK) [reply src/auth/authSlice.ts:52]: `import { localStorage as local } from "redux-persist-webextension-storage";` and then
`storage: local as StorageInterface` ?

I'll second Todd that the purpose of renaming the import, type casting, and "shadowing this type" is not very clear. Could you please elaborate on the reasons?

File: src/pageEditor/store.ts [hunk 6]
```diff
@@ -66,14 +69,14 @@ const store = configureStore({
...
-    editor: editorSlice.reducer,
+    editor: persistReducer(persistEditorConfig, editorSlice.reducer),
     session: sessionSlice.reducer,
     savingExtension: savingExtensionSlice.reducer,
     runtime: runtimeSlice.reducer,
     logs: logSlice.reducer,
     analysis: analysisSlice.reducer,
     tabState: tabStateSlice.reducer,
-    recipes: recipesSlice.reducer,
+    recipes: persistReducer(persistRecipesConfig, recipesSlice.reducer),
```
Comment (BALEHOK) [inline src/pageEditor/store.ts:79]: NIT: I know we've touched this topic, still I can't but mention that persisting an object takes time and degrades editing experience. The less powered is the user's machine the slower is the editing.
We can think of several options from writing performance tests to taking performance reports from client machines to re-designing the way editor updates data in state.

File: src/store/StorageInterface.ts [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+// Exporting from redux-persist-webextension-storage to fix the
+// "exporting variable using name but cannot be named" lint errors
```
Comment (BALEHOK) [inline src/store/StorageInterface.ts:19]: @BLoe do you want to pair on this to try to solve the original lint error?

File: src/store/StorageInterface.ts (hunk 3 above)
Comment (BLoe) [reply src/store/StorageInterface.ts:23]: I was just copying it over from the type in the library. I wasn't sure if there would be an issue changing the type from `any` to `unknown`

File: src/store/StorageInterface.ts (hunk 7 above)
Comment (BLoe) [reply src/store/StorageInterface.ts:19]: It's a type that is private to the `redux-persist-webextension-storage` library, how would we solve it? Happy to take a look tomorrow morning if want to talk about it.

File: src/recipes/recipesSlice.ts (hunk 5 above)
Comment (BLoe) [reply src/recipes/recipesSlice.ts:119]: I think your link is broken there. Is the cache persisted to local storage as well? Happy to remove this for now.

Should we think about using redux-persist in place of our various registries at some point?

File: src/pageEditor/store.ts (hunk 6 above)
Comment (BLoe) [reply src/pageEditor/store.ts:79]: I thought you told me on Slack that we should persist this. I'll remove it.

File: src/auth/authSlice.ts [hunk 8]
```diff
@@ -45,7 +46,10 @@ export const authSlice = createSlice({
 // Current approach is not ideal, AuthState is cached along with UserData (which is used by background script).
 export const persistAuthConfig = {
   key: "authOptions",
-  storage: localStorage,
+  // Change the type of localStorage to our overridden version so that it can be exported
+  // See: @/store/StorageInterface.ts
+  storage: localStorage as StorageInterface,
```
Comment (BALEHOK) [inline src/auth/authSlice.ts:51]: The issue here is that "redux-persist-webextension-storage" exports the variable `localStorage` but doesn't export the interface of this variable `StorageInterface`. As a result TS complaints that it cannot name the type although that name is being used. Copying the interface to our source code and casting is a well acceptable solution.

File: src/store/StorageInterface.ts [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+// Exporting from redux-persist-webextension-storage to fix the
+// "exporting variable using name but cannot be named" lint errors.
+// This type is private in the library.
```
Comment (BALEHOK) [inline src/store/StorageInterface.ts:20]: 👍

File: src/store/extensionsStorage.ts [hunk 10]
```diff
@@ -68,9 +69,12 @@ export async function saveOptions(state: ExtensionOptionsState): Promise<void> {
   });
 }
 
+// Shadowing this type in order to export it properly
+const local: StorageInterface = localStorage;
```
Comment (BALEHOK) [inline src/store/extensionsStorage.ts:73]: please change this as well

**Review by BALEHOK — APPROVED** [review]

**Pull request merged**